crates/biome_line_index/src/lib.rs

```diff
@@ -0,0 +1,171 @@
+//! The crate contains tools for converting between byte offsets and line / column positions.
+
+#![deny(clippy::use_self)]
+
+use biome_text_size::TextSize;
+
+mod line_index;
+
+pub use line_index::LineIndex;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum WideEncoding {
+    Utf16,
+    Utf32,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct LineCol {
+    /// Zero-based
+    pub line: u32,
+    /// Zero-based utf8 offset
+    pub col: u32,
+}
+
+/// Deliberately not a generic type and different from `LineCol`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct WideLineCol {
+    /// Zero-based
+    pub line: u32,
+    /// Zero-based
+    pub col: u32,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct WideChar {
+    /// Start offset of a character inside a line, zero-based
+    pub start: TextSize,
+    /// End offset of a character inside a line, zero-based
+    pub end: TextSize,
+}
+
+impl WideChar {
+    /// Returns the length in 8-bit UTF-8 code units.
+    fn len(&self) -> TextSize {
+        self.end - self.start
+    }
+
+    /// Returns the length in UTF-16 or UTF-32 code units.
+    fn wide_len(&self, enc: WideEncoding) -> usize {
+        match enc {
+            WideEncoding::Utf16 => {
+                if self.len() == TextSize::from(4) {
+                    2
+                } else {
+                    1
+                }
+            }
+
+            WideEncoding::Utf32 => 1,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::WideEncoding::{Utf16, Utf32};
+    use crate::WideLineCol;
+    use crate::line_index::LineIndex;
+    use crate::{LineCol, WideEncoding};
+    use biome_text_size::TextSize;
+
+    macro_rules! check_conversion {
+        ($line_index:ident : $wide_line_col:expr => $text_size:expr ) => {
+            let encoding = WideEncoding::Utf16;
+
+            let line_col = $line_index.to_utf8(encoding, $wide_line_col);
+            let offset = $line_index.offset(line_col);
+            assert_eq!(offset, Some($text_size));
+
+            let line_col = $line_index.line_col(offset.unwrap());
+            let wide_line_col = $line_index.to_wide(encoding, line_col.unwrap());
+            assert_eq!(wide_line_col, Some($wide_line_col));
```

Comment:
This used to use `PositionEncoding::Wide(WideEncoding::Utf16)`, `offset()`, and `position()`, but we don't have access to those here, so I've had to tweak the tests to have the same effect. I think they are spiritually the same, but test at a slightly lower level now.

The original tests with the LSP types still exist in `biome_lsp_converters`